graph/registry_credential.go

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package graph
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+var (
+	errInvalidRegName       = errors.New("registry name can't be empty")
+	errInvalidUsername      = errors.New("username can't be empty")
+	errInvalidPassword      = errors.New("password can't be empty")
+	errInvalidIdentity      = errors.New("identity can't be empty")
+	errInvalidArmResourceID = errors.New("armResourceId can't be empty")
+	errCouldNotClassify     = errors.New("unable to classify credential into opaque, vault or msi")
+)
+
+const (
+	// Opaque means username/password are in plain-text
+	Opaque = "opaque"
+	// VaultSecret means username/password are Azure KeyVault IDs
+	VaultSecret = "vaultsecret"
+)
+
+// RegistryCredential defines a combination of registry, username and password.
+type RegistryCredential struct {
+	Registry      string `json:"registry"`
+	Username      string `json:"username,omitempty"`
+	UsernameType  string `json:"userNameProviderType,omitempty"`
+	Password      string `json:"password,omitempty"`
+	PasswordType  string `json:"passwordProviderType,omitempty"`
+	Identity      string `json:"identity,omitempty"`
+	ArmResourceID string `json:"armResourceId,omitempty"`
+}
+
+// CreateRegistryCredentialFromString creates a RegistryCredential object from a serialized string.
+func CreateRegistryCredentialFromString(str string) (*RegistryCredential, error) {
+	var cred RegistryCredential
+	if err := json.Unmarshal([]byte(str), &cred); err != nil {
+		return nil, errors.Wrap(err, fmt.Sprintf("unable to unmarshal Credentials from string '%s'", str))
```

Comment:
can't return an error with the string here since it might contain sensitive information -- also there's Wrapf